File: core/runtime/src/gpu.rs
//! GPU-dispatch abstraction (ARCHITECTURE §5 "AI accel via MLX / wgpu+tinygrad").
//!
//! A worker that wants to offload an array kernel (vector-add, SAXPY, a matmul
//! tile, …) should not care whether the host has a discrete GPU, an integrated
//! one, or none at all. [`GpuDispatch`] is that seam: submit a [`KernelSource`]
//! plus input buffers, get output buffers back. Two real backends implement it:
//!
//! - [`SoftwareGpu`] — a **real CPU backend**. It actually computes the kernel on
//!   the host (no GPU, no driver), so it is fully testable headlessly and is the
//!   honest fallback on GPU-less machines. Always built.
//! - [`WgpuDispatch`] — a **real wgpu backend** behind the default-OFF `gpu`
//!   feature. It compiles a WGSL compute shader for the same kernel and runs it on
//!   a physical adapter. It *compiles* unconditionally (when the feature is on) but
//!   may legitimately find **no adapter** on a headless/GPU-less host, in which
//!   case [`WgpuDispatch::new`] returns [`GpuError::NoAdapter`]. That is labelled,
//!   not faked: we never pretend a GPU ran when none did (CLAUDE.md "maturity
//!   honesty"). The `gpu` feature is OFF by default so cabi/default builds never
//!   pull `wgpu`.
//!
//! Both backends implement the **same** [`KernelSource`] kernels and must produce
//! identical results — that is what makes the software path a faithful fallback
//! for the GPU path rather than a different computation.

use std::fmt;

/// Errors a [`GpuDispatch`] backend can return.
#[derive(Debug, Clone, PartialEq, Eq)]
pub enum GpuError {
    /// Input buffers had mismatched or unexpected lengths for the kernel.
    BadInputs(String),
    /// No GPU adapter was available (only meaningful for the real GPU backend on a
    /// headless / GPU-less host). The caller should fall back to [`SoftwareGpu`].
    NoAdapter,
    /// The backend failed while compiling/running the kernel.
    Backend(String),
}

impl fmt::Display for GpuError {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            GpuError::BadInputs(m) => write!(f, "bad inputs: {m}"),
            GpuError::NoAdapter => write!(f, "no GPU adapter available"),
            GpuError::Backend(m) => write!(f, "gpu backend error: {m}"),
        }
    }
}

impl std::error::Error for GpuError {}

/// A compute kernel to dispatch. Each variant is an element-wise f32 kernel that
/// **both** backends implement identically (CPU directly; wgpu via a matching WGSL
/// shader), so results are backend-independent.
///
/// These are intentionally small, well-known kernels (the workhorses of a tensor
/// runtime) so the software path can be verified against a hand-computed result
/// and the GPU path checked against the software path.
#[derive(Debug, Clone, Copy, PartialEq)]
pub enum KernelSource {
    /// `out[i] = a[i] + b[i]` over two equal-length input buffers.
    VectorAdd,
    /// `out[i] = alpha * x[i] + y[i]` (BLAS SAXPY) over two equal-length buffers.
    Saxpy { alpha: f32 },
    /// `out[i] = x[i] * scalar` over a single input buffer.
    ScalarMul { scalar: f32 },
}

impl KernelSource {
    /// How many input buffers this kernel consumes.
    fn arity(&self) -> usize {
        match self {
            KernelSource::VectorAdd | KernelSource::Saxpy { .. } => 2,
            KernelSource::ScalarMul { .. } => 1,
        }
    }
}

/// The GPU-dispatch seam: submit a kernel + input buffers, receive output buffers.
///
/// `inputs` are the kernel's input buffers in declared order (see each
/// [`KernelSource`] variant). The return is the single output buffer; its length
/// equals the (common) input length. Implementors must validate arity/length and
/// return [`GpuError::BadInputs`] rather than panicking on a malformed request.
pub trait GpuDispatch {
    /// Run `kernel` over `inputs`, returning the output buffer.
    fn submit(&self, kernel: &KernelSource, inputs: &[&[f32]]) -> Result<Vec<f32>, GpuError>;
}

/// Validate input arity + equal lengths, returning the common length. Shared by
/// both backends so they reject the same malformed requests identically.
fn validate(kernel: &KernelSource, inputs: &[&[f32]]) -> Result<usize, GpuError> {
    let want = kernel.arity();
    if inputs.len() != want {
        return Err(GpuError::BadInputs(format!(
            "kernel needs {want} input buffer(s), got {}",
            inputs.len()
        )));
    }
    let n = inputs[0].len();
    if let Some(bad) = inputs.iter().find(|b| b.len() != n) {
        return Err(GpuError::BadInputs(format!(
            "input buffers must be equal length; expected {n}, got {}",
            bad.len()
        )));
    }
    Ok(n)
}

/// Real **CPU/software** GPU-dispatch backend.
///
/// Computes each [`KernelSource`] directly on the host. No GPU, no driver, fully
/// deterministic and headless — the honest fallback (and the reference the wgpu
/// backend is checked against in tests).
#[derive(Debug, Clone, Copy, Default)]
pub struct SoftwareGpu;

impl SoftwareGpu {
    pub fn new() -> Self {
        Self
    }
}

impl GpuDispatch for SoftwareGpu {
    fn submit(&self, kernel: &KernelSource, inputs: &[&[f32]]) -> Result<Vec<f32>, GpuError> {
        let n = validate(kernel, inputs)?;
        let mut out = vec![0.0f32; n];
        match *kernel {
            KernelSource::VectorAdd => {
                let (a, b) = (inputs[0], inputs[1]);
                for i in 0..n {
                    out[i] = a[i] + b[i];
                }
            }
            KernelSource::Saxpy { alpha } => {
                let (x, y) = (inputs[0], inputs[1]);
                for i in 0..n {
                    out[i] = alpha * x[i] + y[i];
                }
            }
            KernelSource::ScalarMul { scalar } => {
                let x = inputs[0];
                for i in 0..n {
                    out[i] = x[i] * scalar;
                }
            }
        }
        Ok(out)
    }
}

#[cfg(feature = "gpu")]
mod wgpu_backend;
#[cfg(feature = "gpu")]
pub use wgpu_backend::WgpuDispatch;

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn software_vector_add_is_correct() {
        let a = [1.0f32, 2.0, 3.0, 4.0];
        let b = [10.0f32, 20.0, 30.0, 40.0];
        let out = SoftwareGpu::new()
            .submit(&KernelSource::VectorAdd, &[&a, &b])
            .expect("vector add runs");
        assert_eq!(out, vec![11.0, 22.0, 33.0, 44.0]);
    }

    #[test]
    fn software_saxpy_is_correct() {
        // out = 2*x + y
        let x = [1.0f32, 2.0, 3.0];
        let y = [0.5f32, 0.5, 0.5];
        let out = SoftwareGpu
            .submit(&KernelSource::Saxpy { alpha: 2.0 }, &[&x, &y])
            .expect("saxpy runs");
        assert_eq!(out, vec![2.5, 4.5, 6.5]);
    }

    #[test]
    fn software_scalar_mul_is_correct() {
        let x = [1.0f32, -2.0, 3.5];
        let out = SoftwareGpu
            .submit(&KernelSource::ScalarMul { scalar: 3.0 }, &[&x])
            .expect("scalar mul runs");
        assert_eq!(out, vec![3.0, -6.0, 10.5]);
    }

    #[test]
    fn wrong_arity_is_rejected_not_panicked() {
        // VectorAdd needs 2 inputs; give it 1.
        let a = [1.0f32];
        let err = SoftwareGpu.submit(&KernelSource::VectorAdd, &[&a]);
        assert!(matches!(err, Err(GpuError::BadInputs(_))));
    }

    #[test]
    fn mismatched_lengths_are_rejected() {
        let a = [1.0f32, 2.0];
        let b = [1.0f32];
        let err = SoftwareGpu.submit(&KernelSource::VectorAdd, &[&a, &b]);
        assert!(matches!(err, Err(GpuError::BadInputs(_))));
    }

    #[test]
    fn empty_buffers_yield_empty_output() {
        let a: [f32; 0] = [];
        let b: [f32; 0] = [];
        let out = SoftwareGpu
            .submit(&KernelSource::VectorAdd, &[&a, &b])
            .unwrap();
        assert!(out.is_empty());
    }
}

File: daemon/gpu/gpu.go
// Package gpu is the daemon's GPU-dispatch surface (ARCHITECTURE §5 "AI accel via
// wgpu"): submit a small element-wise f32 kernel + input buffers, get the result
// back. It mirrors core/runtime's GpuDispatch (the Rust side) so the two agree
// numerically.
//
// Two backends, selected at build time — and the result ALWAYS reports which one
// actually ran, so nothing is ever presented as "GPU" when a GPU did not run:
//
//   - default build: a REAL pure-Go software backend that computes on the CPU.
//     Works everywhere, needs no C toolchain and no GPU. Reported as "cpu-software".
//   - `-tags ffi`: routes to the Rust core (core/cabi cerberus_gpu_submit), which
//     runs the physical GPU via wgpu when cabi is built `--features gpu` and an
//     adapter is present ("gpu-wgpu"), else the Rust software backend. On any FFI
//     error it falls back to the pure-Go backend so a real result still comes back.
//
// A kernel never fabricates a result: an unsupported/invalid request errors.
package gpu

import "fmt"

// Kernel identifies an element-wise f32 kernel. The integer values are the ABI
// contract with core/cabi's cerberus_gpu_submit (0/1/2) — do not renumber.
type Kernel int

const (
	VectorAdd Kernel = 0 // out[i] = a[i] + b[i]
	Saxpy     Kernel = 1 // out[i] = alpha*a[i] + b[i]  (param = alpha)
	ScalarMul Kernel = 2 // out[i] = a[i] * scalar       (param = scalar; b ignored)
)

func (k Kernel) String() string {
	switch k {
	case VectorAdd:
		return "vector-add"
	case Saxpy:
		return "saxpy"
	case ScalarMul:
		return "scalar-mul"
	default:
		return fmt.Sprintf("kernel(%d)", int(k))
	}
}

// ParseKernel maps a CLI kernel name to a Kernel.
func ParseKernel(name string) (Kernel, bool) {
	switch name {
	case "vector-add", "vectoradd", "add":
		return VectorAdd, true
	case "saxpy":
		return Saxpy, true
	case "scalar-mul", "scalarmul", "scale":
		return ScalarMul, true
	default:
		return 0, false
	}
}

// arity is how many input buffers the kernel consumes.
func (k Kernel) arity() int {
	if k == ScalarMul {
		return 1
	}
	return 2
}

// validate rejects unknown kernels and mismatched input lengths (the same checks
// core/runtime's validate() makes, so both sides reject identically).
func validate(k Kernel, a, b []float32) error {
	if k < VectorAdd || k > ScalarMul {
		return fmt.Errorf("unknown kernel %d", int(k))
	}
	if k.arity() == 2 && len(a) != len(b) {
		return fmt.Errorf("%s needs equal-length inputs, got %d and %d", k, len(a), len(b))
	}
	return nil
}

// software runs the kernel on the CPU in pure Go — a real reference backend,
// numerically identical to core/runtime's SoftwareGpu (and its WGSL shaders). It
// is the default backend and the fallback for the FFI path.
func software(k Kernel, param float32, a, b []float32) ([]float32, error) {
	if err := validate(k, a, b); err != nil {
		return nil, err
	}
	n := len(a)
	out := make([]float32, n)
	switch k {
	case VectorAdd:
		for i := 0; i < n; i++ {
			out[i] = a[i] + b[i]
		}
	case Saxpy:
		for i := 0; i < n; i++ {
			out[i] = param*a[i] + b[i]
		}
	case ScalarMul:
		for i := 0; i < n; i++ {
			out[i] = a[i] * param
		}
	}
	return out, nil
}

// Dispatch runs kernel k over the input buffer(s) on the best available backend
// and returns the result plus the name of the backend that ACTUALLY ran it
// ("cpu-software", or "gpu-wgpu"/"rust-software" under -tags ffi). b is ignored
// for ScalarMul. It is the single entry point the RPC/CLI call.
func Dispatch(k Kernel, param float32, a, b []float32) (out []float32, backend string, err error) {
	return dispatch(k, param, a, b)
}
